ws4/02_Unit-Tests.ipynb

```diff
@@ -432,6 +438,324 @@
+    "- [ ] A new server has 2 installs after 2 calls to `install`:\n",
+    "  - **Given**: A new instance of `Server`,\n",
+    "  - **When**: `server.install(\"firefox\")` and `server.install(\"python\")` is called,\n",
+    "  - **Then**: The instance has `\"firefox\"` and `\"python\"` in `installed`."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "editable": true,
+    "slideshow": {
+     "slide_type": "notes"
+    },
+    "tags": []
+   },
+   "source": [
+    "## Solution (Exercise 1)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "editable": true,
+    "slideshow": {
+     "slide_type": "notes"
+    },
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "import server"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "editable": true,
+    "slideshow": {
+     "slide_type": "notes"
+    },
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "%%pytest\n",
+    "\n",
+    "import pytest\n",
+    "\n",
+    "from server import Server\n",
+    "\n",
+    "def test_new_server_has_no_installed_software():\n",
+    "    server = Server()\n",
+    "    assert not server.installed\n",
+    "\n",
+    "def test_new_server_has_one_install_after_call_to_install():\n",
+    "    server = Server()\n",
+    "    server.install(\"firefox\")\n",
+    "    assert server.installed == [\"firefox\"]\n",
+    "\n",
+    "def test_new_server_has_2_installs_after_2_calls_to_install():\n",
+    "    server = Server()\n",
+    "    server.install(\"firefox\")\n",
+    "    server.install(\"python\")\n",
+    "    assert set(server.installed) == {\"firefox\", \"python\"}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "editable": true,
+    "slideshow": {
+     "slide_type": "subslide"
+    },
+    "tags": []
+   },
+   "source": [
+    "## Exercise 2\n",
+    "We will test a modified version of command line interface `iam_example.py` from section 3.\n",
+    "Note the following change, which makes the command line interface testable, by adding an `*args` argument to `main()` and passing it on to `parser.parse_args()`:\n",
+    "\n",
+    "```python\n",
+    "def main(*args):\n",
+    "    ...\n",
+    "    args = parser.parse_args(*args)\n",
+    "    ...\n",
+    "```\n",
+    "\n",
+    "This will allow us to call for example\n",
+    "```python\n",
+    "main([\"user\", \"--add\", \"chucknorris\"])\n",
+    "```\n",
+    "Which corresponds to the following prompt from the command line:\n",
+    "```sh\n",
+    "./iam_example.py user --add chucknorris\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "editable": true,
+    "slideshow": {
+     "slide_type": "subslide"
+    },
+    "tags": []
+   },
+   "source": [
+    "This change allows us to pass command line arguments directly to `main` within python."
+   ]
+  },
```

Comment:
can be deleted if you apply my suggestion above.

Reply:
Done.
Thanks a lot @pulsargranular for your very valuable additions!